extension/src/extension.ts

```diff
@@ -33,39 +34,39 @@ export function activate(context: vscode.ExtensionContext) {
     "gdb_script.py",
   );
 
-  const trackerDisposable = vscode.debug.registerDebugAdapterTrackerFactory(
-    "cppdbg",
-    {
-      createDebugAdapterTracker(session: vscode.DebugSession) {
-        if (!settings.enabled) return undefined;
-        return {
-          onWillReceiveMessage: async (message: DebugProtocol.Request) => {
-            if (handler === null) {
-              return;
-            }
-            // TODO: request queue?
-            await handler.handleMessageFromClient(message);
-          },
-          onDidSendMessage: (
-            message: DebugProtocol.Event | DebugProtocol.Response,
-          ) => {
-            if (handler === null) {
-              return;
-            }
-            handler.handleMessageToClient(message);
-          },
-        };
+  const messageQueue = new MessageQueue();
+
+  function createTrackerFactory(debuggerType: string): vscode.Disposable {
+    return vscode.debug.registerDebugAdapterTrackerFactory(
+      debuggerType,
+      {
+        createDebugAdapterTracker(session: vscode.DebugSession) {
+          if (!settings.enabled) return undefined;
+          return {
+            onWillReceiveMessage: async (message: DebugProtocol.Request) => {
+              messageQueue.enqueue(message, MessageType.Incoming);
```

Comment:
I'm not sure if the `await` here wasn't load-bearing. But as long as we don't ever send more than a single item from the queue to the handler, and make sure that they are processed in order, hopefully it will be fine.